### PR TITLE
[LOG4J2-3341] Support logger level and appender refs shorthand

### PR DESCRIPTION
**Pull request by jvz**
This adds a properties format feature where the property names "rootLogger" and "logger.<key>" can be set to a level and one or more appender refs.

https://issues.apache.org/jira/browse/LOG4J2-3341

### CODE REVIEW

File: log4j-core/src/test/resources/LoggerLevelAppenderTest.properties
Comment (Apache9) [inline log4j-core/src/test/resources/LoggerLevelAppenderTest.properties:21]: Good. I think this is enough for hadoop and hbase, although it maybe a pain that we need to maintain the index. It will be easier to specify appender.first.type = List. But anyway, it works.

Another question is how to specify leven and appenders at once in xml configuration file?

Comment (jvz) [reply log4j-core/src/test/resources/LoggerLevelAppenderTest.properties:21]: The indexed bit is how the current properties format works. I think a separate feature could be added here to automatically set the `name` attribute to the property key when otherwise absent.

Comment (rgoers) [reply log4j-core/src/test/resources/LoggerLevelAppenderTest.properties:21]: Matt and I only worked out how to do it for properties as that was what was requested. Doing it for XML could be possible, but it would be more difficult in YAML and JSON. I'm not even sure what the point of that would be.

For XML the syntax would have to look like

<Logger name="org.apache.logging.log4j.core.config.properties" level="INFO" appenderRefs="first,second"/>

That is actually not "normal" for XML to have a list specified that way. And there would definitely be no reason to combine the level and appenderRefs into a single attribute.
Also, looking at the Log4j1 DTD the xml had to be specified as 

```
  <root>
    <priority value ="trace" />
    <appender-ref ref="A1" />
    <appender-ref ref="A2" />
  </root>
```

Note that the priority (which could also be specified as level) had to be an XML element whereas Log4j 2 allows it to be an element or an attribute. 
So the bottom line is, I don't know why we would add the support to XML since it doesn't buy anything in the way of compatibility with Log4j 1.

Comment (jvz) [reply log4j-core/src/test/resources/LoggerLevelAppenderTest.properties:21]: I've added a way to do implicit appender names, too. As for the XML config variant, that would need to add the CompoundProperty plugin.

File: log4j-core/src/test/resources/RootLoggerLevelAppenderTest.properties
Comment (jvz) [inline log4j-core/src/test/resources/RootLoggerLevelAppenderTest.properties:21]: Now this properties format is getting less shitty to use, too! I like this.

**Review by rgoers — APPROVED** [review]

**Comment by ppkarwasz** [comment]
@jvz: maybe you can add a default for all component names: `name = key.replaceAll("-", ".")`? This way you'll allow also to configure a logger in one property:

`logger.org-apache-logging-log4j=INFO,app1,app2`

File: log4j-core/src/test/resources/LoggerLevelAppenderTest.properties
Comment (Apache9) [reply log4j-core/src/test/resources/LoggerLevelAppenderTest.properties]: OK, got it. So this is a properties file specific way.

**Review by Apache9 — APPROVED** [review]

File: log4j-core/src/main/java/org/apache/logging/log4j/core/config/properties/PropertiesConfigurationBuilder.java
Comment (carterkozak) [inline log4j-core/src/main/java/org/apache/logging/log4j/core/config/properties/PropertiesConfigurationBuilder.java:229]: I think we may be slightly better off retaining the original `remove` logic, only traversing the map once:
```suggestion
        final String name = Objects.toString(properties.remove(CONFIG_NAME), key);
```

File: log4j-core/src/main/java/org/apache/logging/log4j/core/config/properties/PropertiesConfigurationBuilder.java
Comment (carterkozak) [inline log4j-core/src/main/java/org/apache/logging/log4j/core/config/properties/PropertiesConfigurationBuilder.java:200]: We should check that `parts` is not empty to avoid an `ArrayIndexOutOfBoundsException`. If memory serves, `",".split(",")` will result in an empty array.

File: log4j-core/src/main/java/org/apache/logging/log4j/core/config/properties/PropertiesConfigurationBuilder.java
Comment (carterkozak) [inline log4j-core/src/main/java/org/apache/logging/log4j/core/config/properties/PropertiesConfigurationBuilder.java:202]: It may be worthwhile to add some StatusLogger debug logging here, I defer to your judgement :-)

File: log4j-core/src/main/java/org/apache/logging/log4j/core/config/properties/PropertiesConfigurationBuilder.java
Comment (carterkozak) [inline log4j-core/src/main/java/org/apache/logging/log4j/core/config/properties/PropertiesConfigurationBuilder.java:199]: Do we need to trim each segment of `parts` to avoid whitespace issues? I suspect the level piece may handle it gracefully, but I'm less sure about the appenderRefs. Could be worth a test.

**Review by carterkozak — COMMENTED** [review]
Very nifty!

File: log4j-core/src/main/java/org/apache/logging/log4j/core/config/properties/PropertiesConfigurationBuilder.java
Comment (jvz) [reply log4j-core/src/main/java/org/apache/logging/log4j/core/config/properties/PropertiesConfigurationBuilder.java:200]: I think that split would result in a two value array of two empty strings.

**Comment by jvz** [comment]
> @jvz: maybe you can add a default for all component names: `name = key.replaceAll("-", ".")`? This way you'll allow also to configure a logger in one property:
> 
> `logger.org-apache-logging-log4j=INFO,app1,app2`

This is an interesting idea. The only potential issue here is that dashes are technically allowed in logger names, though I doubt they're used. Could probably be added with this whole update.

I'll have docs updates to go with this before it's ready to merge.

File: log4j-core/src/main/java/org/apache/logging/log4j/core/config/properties/PropertiesConfigurationBuilder.java
Comment (jvz) [reply log4j-core/src/main/java/org/apache/logging/log4j/core/config/properties/PropertiesConfigurationBuilder.java:199]: Yes, that's a good point.

- [ ] Trim appender refs

File: log4j-core/src/test/resources/RootLoggerLevelAppenderTest.properties
Comment (jvz) [reply log4j-core/src/test/resources/RootLoggerLevelAppenderTest.properties:21]: Updated.

File: log4j-core/src/main/java/org/apache/logging/log4j/core/config/properties/PropertiesConfigurationBuilder.java
Comment (jvz) [reply log4j-core/src/main/java/org/apache/logging/log4j/core/config/properties/PropertiesConfigurationBuilder.java]: Good idea; added.

File: log4j-core/src/main/java/org/apache/logging/log4j/core/config/properties/PropertiesConfigurationBuilder.java
Comment (carterkozak) [reply log4j-core/src/main/java/org/apache/logging/log4j/core/config/properties/PropertiesConfigurationBuilder.java]: It's subtle and counter-intuitive. Alas, this test passes:
```java
@Test
public void split() {
    assertThat(",".split(",")).isEmpty();
}
```

**Review by carterkozak — APPROVED** [review]
Looks good to me. Thanks, Matt!

File: log4j-core/src/main/java/org/apache/logging/log4j/core/config/properties/PropertiesConfigurationBuilder.java
Comment (jvz) [reply log4j-core/src/main/java/org/apache/logging/log4j/core/config/properties/PropertiesConfigurationBuilder.java]: Indeed it is! I only remembered that because I checked the javadocs for `split` before using it here. At first, I was limiting the split to 1 match, but then I added the multiple refs thing.

**Comment by jvz** [comment]
I've squashed what's been reviewed. Will add more commits for docs and changelog.

**Comment by ppkarwasz** [comment]
@jvz: #729 drops previously invalid entries like `logger.core-config-properties=INFO,app1,app2`. You should probably add a test for that.

**Comment by jvz** [comment]
I have some code in PropertiesConfigurationBuilder that manually includes the invalid keys left behind.

**Pull request merged**
